src/ServiceControl/SagaAudit/SagaUpdatedHandler.cs

```diff
@@ -1,24 +1,102 @@
 ﻿namespace ServiceControl.SagaAudit
 {
+    using System;
+    using System.Collections.Concurrent;
+    using System.Threading;
     using System.Threading.Tasks;
     using EndpointPlugin.Messages.SagaState;
+    using Newtonsoft.Json.Linq;
     using NServiceBus;
-    using ServiceControl.Persistence;
+    using ServiceControl.Connection;
+    using ServiceControl.Infrastructure.DomainEvents;
 
     class SagaUpdatedHandler : IHandleMessages<SagaUpdatedMessage>
     {
-        public SagaUpdatedHandler(ISagaAuditDataStore store)
+        public SagaUpdatedHandler(IDomainEvents domainEvents, IPlatformConnectionBuilder connectionBuilder)
         {
-            this.store = store;
+            this.domainEvents = domainEvents;
+            this.connectionBuilder = connectionBuilder;
         }
 
-        public Task Handle(SagaUpdatedMessage message, IMessageHandlerContext context)
+        public async Task Handle(SagaUpdatedMessage message, IMessageHandlerContext context)
         {
-            var sagaSnapshot = SagaSnapshotFactory.Create(message);
+            await WarnOncePerEndpointPerDay(message);
 
-            return store.StoreSnapshot(sagaSnapshot);
+            if (auditQueueName is null || nextAuditQueueNameRefresh < DateTime.UtcNow)
+            {
+                await RefreshAuditQueue();
+            }
+
+            if (auditQueueName is null)
+            {
+                throw new Exception("Could not determine audit queue name to forward saga update message. The ServiceControl remote audit instance ");
+            }
+
+            await context.ForwardCurrentMessageTo(auditQueueName);
+        }
+
+        async Task RefreshAuditQueue()
+        {
+            await semaphore.WaitAsync();
+            try
+            {
+                if (auditQueueName != null && nextAuditQueueNameRefresh > DateTime.UtcNow)
+                {
+                    return;
+                }
+
+                var connectionDetails = await connectionBuilder.BuildPlatformConnection();
+
+                if (connectionDetails.ToDictionary().TryGetValue("SagaAudit", out var sagaAuditObj) && sagaAuditObj is JObject sagaAudit)
```

Comment:
How does this behave when there are multiple audit instances?

Reply:
Here is where each audit instance would respond with its own data:

https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl.Audit/Connection/ConnectionController.cs#L16-L31

Here's where the primary would be asking each instance for that:

https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl/Connection/RemotePlatformConnectionDetailsProvider.cs#L29-L57

On line 46 you can see it's calling `connection.Add(…)` for each key it gets back which is here:

https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl/Connection/PlatformConnectionDetails.cs#L10-L23

So if you have multiple audit instances all providing a `SagaAudit` key, the rest will become `SagaAudit1`, `SagaAudit2`, etc. but we'll take the value from the first non-numbered one.

And ultimately they should all be the same queue anyway. Unless you're on MSMQ and the Audit instance doesn't know about servernames. But if that's the case then the Platform Connection feature has a bug in it. In any case I intend to check that later on.